pylabrobot/resources/carrier.py

```diff
@@ -251,21 +251,6 @@ def __init__(
       sites,category=category, model=model)
 
 
-class ShakerCarrier(Carrier):
-  """ Base class for shaker carriers (i.e. 7-track carriers with mixed-use and/or specialized
-  sites). """
-  def __init__(
-    self,
-    name: str,
-    size_x: float,
-    size_y: float,
-    size_z: float,
-    sites: Optional[List[CarrierSite]] = None,
-    category="shaker_carrier",
-    model: Optional[str] = None):
-    super().__init__(name, size_x, size_y, size_z,
-      sites,category=category, model=model)
-
```

Comment:
why did you remove this?

Reply:
Because there are no `ShakerCarrier`s:
There is only one 7-track carrier that can flexibly take different modules onto it, including shakers, coolers, plate stack sites, NTR sites, etc.
I have already added this specific 7-track carrier called `PLT_CAR_L4_SHAKER` as a `MFXModule`.

As a result this `ShakerCarrier` class has never been used and, to my knowledge, has no opportunity to be used in the future.

I created this class before I understood that `PLT_CAR_L4_SHAKER` is better represented as a `MFXModule`.

Reply:
That makes sense. With plr maturing and starting to be used by multiple different parties, I would encourage breaking changes be introduced more gently. For example with this one we could do something like:

```
class ShakerCarrier(PLT_CAR_L4_SHAKER ):
  """ Deprecated -  use PLT_CAR_L4_SHAKER  """
  def __init__(
    self,
    name: str,
    size_x: float,
    size_y: float,
    size_z: float,
    sites: Optional[List[CarrierSite]] = None,
    category="shaker_carrier",
    model: Optional[str] = None):
    # TODO #<number>: On August 31, 2024, make this raise an exception with the warn message.
    logger.warn("ShakerCarrier has been deprecated in favor of PLT_CAR_L4_SHAKER ")
    super().__init__(name, size_x, size_y, size_z,
      sites,category=category, model=model)
```
this would be backwards compatible and also alert people that they should update their projects.

Reply:
Agreed, we should communicate breaking changes in a gentler way. I will watch this in PRs. We should discuss versioning and a deprecation schedule soon-ish. I will set up that discussion.

@jkhales, to my knowledge, Retro is the only one using the ShakerCarrier class. Do you think you could update this now? then we can avoid having this warning and unused carrier.

Reply:
yea, np. But we also don't know who else is using this stuff, there could be some quiet person who just set it up in a small midwestern lab for whom it breaks and they don't know what to do and just give up on plr because it's so frustrating.

The cost of adding the message is very cheap. At the very least perhaps make it raise an Exception with directions on what to do? Could save someone hours.

Reply:
I don't understand: How is this a breaking change, there is no instance of a `ShakerCarrier` in the PLR Resource Library?

I made this class in anticipation that I will add one but when I contributed `PLT_CAR_L4_SHAKER` I did so straight away as a `MFXCarrier`.
PLR can only see what is in PLR. If people make their own definition and don't contribute them, how would PLR ever know about it?

Reply:
> I don't understand: How is this a breaking change, there is no instance of a ShakerCarrier in the PLR Resource Library?

if someone is using it (me) in the code and it disappears, it breaks all of the code. This means it is "breaking". 

I'm unfamiliar with any sort of special repository that is "experimental" or  conversely "stable", i believe @rickwierenga shared this with me as something i can use. Is this the case?

Reply:
> if someone is using it (me) in the code and it disappears, it breaks all of the code. This means it is "breaking".

I think I understand now: I contributed `ShakerCarrier` in [PR#59 ](https://github.com/PyLabRobot/pylabrobot/pull/59/files) (2024-01-14), and then  contributed the only Hamilton shaker carrier `PLT_CAR_L4_SHAKER` in [PR#130](https://github.com/PyLabRobot/pylabrobot/pull/130/files) (2024-05-15).
At some point in the 4 months in between you created `PLT_CAR_L4_SHAKER` but as a `ShakerCarrier` (not a `MFXCarrier`) instance yourself and didn't contribute it to PLR?

I'm referring to "breaking" in the domain of PyLabRobot. Nobody can see into what people privately do with PLR.
But this cannot be included in the development of PLR decision making because nobody knows about it except for the person that does not contribute.

Of course, I have a lot of privately defined resources as well.
A number of PRs/commits "broke" these on several occasions with new developments of PLR:
e.g. the recent switch from `pylabrobot.resources.create_equally_spaced` to `pylabrobot.resources.utils.create_equally_spaced_2d` (commit [d16d995](https://github.com/PyLabRobot/pylabrobot/commit/d16d995b01ab6b6eda51b323271bdea5a11b953a?diff=unified&w=0)) "broke" all my private, non-contributed plate definitions.
But I would not expect the contributor to know about this, and stop or delay development of PLR because they didn't know about my non-contributed plate definitions.
The contributor, in this case @rickwierenga, did go through the necessary due diligence and changed all instances of `create_equally_spaced` to `create_equally_spaced_2d` **inside PLR**, i.e. there was no "break" of any functionality for users who use PLR contributed plates.

Similarly, I went through the PLR Resource Library and found no instance of `ShakerCarrier`.
Together with the open discussion of the developing [PLR resource naming convention](https://labautomation.io/t/standardized-labware-naming-convention-discussion/3787/5), in which nobody mentioned the need for an extra `ShakerCarrier` that can only define one carrier that has already been defined, I therefore removed my unused contribution to clean up the Resource Library.

Reply:
I'm using it in a script directly:

<img width="665" alt="image" src="https://github.com/PyLabRobot/pylabrobot/assets/1220444/af3fb796-1749-4142-87d7-6feec65324b4">

> e.g. the recent switch from pylabrobot.resources.create_equally_spaced to pylabrobot.resources.utils.create_equally_spaced_2d (commit [d16d995](https://github.com/PyLabRobot/pylabrobot/commit/d16d995b01ab6b6eda51b323271bdea5a11b953a?diff=unified&w=0)) "broke" all my private, non-contributed plate definitions.

yea, so the way I would suggest doing this since plr is maturing is to follow a more standard approach. One aggressive way would be to do:

```
async def create_equally_spaced(...):
    raise Exception("deprecate - use create_equally_spaced_2d instead.")
```

this would at least allow people to know how to fix things quickly.

A more standard way would be to do:

async def create_equally_spaced(...):
    logger.warning("deprecate - use create_equally_spaced_2d instead.")
    return await create_equally_spaced_2d(...)
```
and then give people a few months to update.

>  stop or delay development of PLR
this doesn't stop or delay! it's just documenting the changes in a user friendly manner.

> But I would not expect the contributor to know about this

If plr is going to become a *mature* library that is trusted and used by people, then it will have to start thinking about backwards compatibility.  It's never going to be taken seriously if breaking changes are pushed constantly without any communication or thought. I know plr is in a rapid development phase, but I'm trying to phase in slowly and get people thinking about how a professional OSS project is maintained.

Reply:
> yea, so the way I would suggest doing this since plr is maturing is to follow a more standard approach. One aggressive way would be to do:
> 
> ```
> async def create_equally_spaced(...):
>     raise Exception("deprecate - use create_equally_spaced_2d instead.")
> ```
> 
> this would at least allow people to know how to fix things quickly.
> 
> A more standard way would be to do:
> 
> ```
> async def create_equally_spaced(...):
>     logger.warning("deprecate - use create_equally_spaced_2d instead.")
>     return await create_equally_spaced_2d(...)
> ```
> 
> and then give people a few months to update.

Okay, this convinced me. Let's try and implement this with all major changes we make and see how we can avoid creating a list of timers of warning triggers that we then have to manually maintain with the handful of people that actively develop PLR.
But I can see and share your point about backward compatibility, even though we are talking about the development branch, i.e. PLR:main, and everyone should still check what changes they are pulling into their private production branches when syncing to the main branch anyway.
In the end, people who don't contribute their resource definitions will still end up having to manually sync their non-contributed resources to PLR's changes, they just get a couple of weeks to do it before breaking - if they chose to sync to the developer's branch latest updates.

I think we should have a small developer meeting in the next 1-2 weeks to discuss and write down how to tackle this, and answer questions pertaining to its implementation, e.g. how long to keep warning triggers before removing them (e.g. 2 months), and potentially how to automate their removal?

In this light, a warning note: last week Rick and I updated the lid definition. Rick went through all PLR-hosted plate definitions and updated them. But all of my privately-hosted plate definitions broke after I synced to the latest developer version on PLR:main and I had to manually update them.